### PR TITLE
Add an option to allow adding Kubernetes Labels as Prometheus Labels per metric series

### LINKED ISSUE #1415: An option to add Kuberntes labels as Prometheus labels to metrics
**What would you like to be added**:

To be able to add allow-listed Kubernetes labels as Prometheus labels directly to certain metrics.

**Why is this needed**:

`kube-state-metrics` exposes Kuberntes roles and labels via special metrics (`kube_node_labels`, `kube_node_role`, `kube_pod_labels`, etc) As a result, when we need to query information from a subset of nodes that that belongs role or label, we need to join the corresponding metrics with `kube_node_labels` on the client side (if the client side supports such operation) which make the query more complex and error prone.

For example,
```
kube_node_status_condition{condition="Ready", status="true"} == 0) and on (node) (kube_node_labels{label_nvidia_com_gpu_product!=""}
```

**Describe the solution you'd like**

It would be great to be able to write
```
kube_node_status_condition{condition="Ready", status="true", label_nvidia_com_gpu_product!=""} == 0
```
instead of
```
kube_node_status_condition{condition="Ready", status="true"} == 0) and on (node) (kube_node_labels{label_nvidia_com_gpu_product!=""}
```

**Additional context**

### LINKED ISSUE #1758: Add an option to allow adding Kubernetes Labels as Prometheus Metric Labels per metric-series
I created this Issue for discussing and tracking PR #1689 
Below is a transcript of the PR 👇🏻
------

# What this PR does / why we need it:

This PR Introduce a new flag `--per-metric-labels-allowlist` that its syntax works similar to `--metric-labels-allowlist` but instead of being a filter to add labels to `kube_X_labels`, it is a filter for K8S' labels that will be added to **each metric** time series of a resource.

## Motivation

Motivation for this change can be better described from a Platform Team POV, who is responsible for the observability stack and provides it to multiple teams and/or multiple tenants.

**The goal is to make it easier to define queries, alerts, and rules without the need for complex joins. Making the barrier for smaller less experienced teams smaller.** As well as alleviate pressure from Prometheus Server that's constantly doing joins for each alert rule.

#### Use Case 1
1. A Development Team wants to create alerts for the multiple components of their applications.
2. **Different Components have different alerts, severities, thresholds**. (example, web pods, background consumers, different kinds of jobs) and as components live in the same namespace; filtering with a namespace is not feasible.
3. You'll now have to use joins with `kube_X_labels` to filter for the specific resources. Complex queries become more complex, especially ones that had joins already.

#### Use Case 2
1. Platform Team defining **general, default rules** for every namespace.
2. Platform team expects resources to have a set of standard labels, that defines something like `alerting.xxx/severity`, `alerting.xxx/slack-channel` along side the `app.kubernetes.io/name`, `app.kubernetes.io/component` ones.
3. Complex Rules are defined to join with these labels and generate alerts with variables based on labels defined by teams. Queries become even more complex as you want to join with **more than one label**.

### Complex Queries Example

`Deployment has not matched the expected number of replicas.`

- Using only 1 label for joins

   ```prometheus
  (
    kube_deployment_spec_replicas{} * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
    !=
    kube_deployment_status_replicas_available{} * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{}[10m] ) * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
    ==  0
  )
   ```

- Using 2 label for joins
   ```prometheus
  (
    kube_deployment_spec_replicas{} * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
    !=
    kube_deployment_status_replicas_available{} * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{}[10m] ) * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
    ==  0
  )
  ```

Same query but with labels as part of metric series:

- ```
  (
    kube_deployment_spec_replicas{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"} 
    !=
    kube_deployment_status_replicas_available{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}[10m] )
  )
  ```
- ```
  (
    kube_deployment_spec_replicas{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"} 
    !=
    kube_deployment_status_replicas_available{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}[10m] )
  )
  ```

## Goal
- Making it easier for Platform and Development teams to create queries, rules, alerts, and dashboards for resources running in Kubernetes without the need for complex joins to filter resources.
- Alleviate some pressure from the Prometheus Servers that are constantly running Joins for each alert rule.

### Alternatives we tried
- Run recording rules to pre-compute the joins:
Having to have a recording rule **for each metric series** generated by KSM is cumbersome, and adds a dependency between teams and platform teams to add any recording rule. Not ideal especially in a multi-tenant environment.

# How does this change affect the cardinality of KSM
*(increases, decreases, or does not change cardinality)*

- Unless explicitly using the new flag, this PR has no change to generated metrics cardinality.
- Using the flag will add new labels for each metric series of each resource that has a label key that's whitelisted.
- Cardinality of the label values depends on how often these labels change for the same resource.
- For the use-case behind this new feature, whitelisted labels are often labels that don't change often. Admins should be cautious of what labels to whitelist.

### Performance
- KSM already fetches each resource label during metric collection.
- Prometheus Performance shouldn't be affected as long as whitelisted labels are not changing constantly.

# Misc

**Which issue(s) this PR fixes**:
- Fixes #1415

**Relevant**:
- [Dynamic alert routing with Prometheus and Alertmanager](https://tech.loveholidays.com/dynamic-alert-routing-with-prometheus-and-alertmanager-f6a919edb5f8)

**Notes**:
- ⚠️ This is a Draft PR, the implementation is not final, PR is a working POC for `pods` resource, and is yet to be discussed.

### PR DESCRIPTION
**Pull request by sherifabdlnaby**
# What this PR does / why we need it:

This PR Introduce a new flag `--per-metric-labels-allowlist` that its syntax works similar to `--metric-labels-allowlist` but instead of being a filter to add labels to `kube_X_labels`, it is a filter for K8S' labels that will be added to **each metric** time series of a resource.

## Motivation

Motivation for this change can be better described from a Platform Team POV, who is responsible for the observability stack and provides it to multiple teams and/or multiple tenants.

**The goal is to make it easier to define queries, alerts, and rules without the need for complex joins. Making the barrier for smaller less experienced teams smaller.** As well as alleviate pressure from Prometheus Server that's constantly doing joins for each alert rule.

#### Use Case 1
1. A Development Team wants to create alerts for the multiple components of their applications.
2. **Different Components have different alerts, severities, thresholds**. (example, web pods, background consumers, different kinds of jobs) and as components live in the same namespace; filtering with a namespace is not feasible.
3. You'll now have to use joins with `kube_X_labels` to filter for the specific resources. Complex queries become more complex, especially ones that had joins already.

#### Use Case 2
1. Platform Team defining **general, default rules** for every namespace.
2. Platform team expects resources to have a set of standard labels, that defines something like `alerting.xxx/severity`, `alerting.xxx/slack-channel` along side the `app.kubernetes.io/name`, `app.kubernetes.io/component` ones.
3. Complex Rules are defined to join with these labels and generate alerts with variables based on labels defined by teams. Queries become even more complex as you want to join with **more than one label**.

### Complex Queries Example

`Deployment has not matched the expected number of replicas.`

- Using only 1 label for joins

   ```prometheus
  (
    kube_deployment_spec_replicas{} * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
    !=
    kube_deployment_status_replicas_available{} * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{}[10m] ) * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
    ==  0
  )
   ```

- Using 2 label for joins
   ```prometheus
  (
    kube_deployment_spec_replicas{} * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
    !=
    kube_deployment_status_replicas_available{} * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{}[10m] ) * on (deployment) group_left(label_app_kubernetes_io_name, label_alerting_severity) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
    ==  0
  )
  ```

Same query but with labels as part of metric series:

- ```
  (
    kube_deployment_spec_replicas{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"} 
    !=
    kube_deployment_status_replicas_available{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}[10m] )
  )
  ```
- ```
  (
    kube_deployment_spec_replicas{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"} 
    !=
    kube_deployment_status_replicas_available{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}
  ) 
  and 
  (
    changes(kube_deployment_status_replicas_updated{label_app_kubernetes_io_name="emojiapp", label_alerting_severity="critical", namespace=~"emoji"}[10m] )
  )
  ```

## Goal
- Making it easier for Platform and Development teams to create queries, rules, alerts, and dashboards for resources running in Kubernetes without the need for complex joins to filter resources.
- Alleviate some pressure from the Prometheus Servers that are constantly running Joins for each alert rule.

### Alternatives we tried
- Run recording rules to pre-compute the joins:
Having to have a recording rule **for each metric series** generated by KSM is cumbersome, and adds a dependency between teams and platform teams to add any recording rule. Not ideal especially in a multi-tenant environment.

# How does this change affect the cardinality of KSM
*(increases, decreases, or does not change cardinality)*

- Unless explicitly using the new flag, this PR has no change to generated metrics cardinality.
- Using the flag will add new labels for each metric series of each resource that has a label key that's whitelisted.
- Cardinality of the label values depends on how often these labels change for the same resource.
- For the use-case behind this new feature, whitelisted labels are often labels that don't change often. Admins should be cautious of what labels to whitelist.

### Performance
- KSM already fetches each resource label during metric collection.
- Prometheus Performance shouldn't be affected as long as whitelisted labels are not changing constantly.

# Misc

**Which issue(s) this PR fixes**:
- Fixes #1415
- Fixes #1758

**Relevant**:
- [Dynamic alert routing with Prometheus and Alertmanager](https://tech.loveholidays.com/dynamic-alert-routing-with-prometheus-and-alertmanager-f6a919edb5f8)

**Notes**:
- ⚠️ This is a Draft PR, the implementation is not final, PR is a working POC for `pods` resource, and is yet to be discussed.

### CODE REVIEW

File: internal/store/builder.go [hunk 1]
```diff
@@ -205,6 +207,13 @@ func (b *Builder) WithAllowLabels(labels map[string][]string) {
 	}
 }
 
+// WithAllowMetricLabels configures which labels can be returned for each metric
+func (b *Builder) WithAllowMetricLabels(labels map[string][]string) {
+	if len(labels) > 0 {
+		allowMetricLabelsList = labels
```
Comment (sherifabdlnaby) [inline internal/store/builder.go:213]: For POC purposes I used a Package global value to be available to the `wrapPodFunc()` without having to pass it down multiple layers. This was I avoided changing the signature of multiple functions.

KSM doesn't use global variables to share config, however, to implement this across all resources we'll need to change the signature of every `create<Resource><Metric>FamilyGenerator` function of each resource and pass the `allowMetricLabelsList`.

**Comment by sherifabdlnaby** [comment]
@dmitri-lerko Has an article about building a platform for teams on top of KSM, I'd like if you'd join the conversation here 👋🏻

**Comment by fpetkovski** [comment]
Thanks for this contribution. If I understand correctly, this PR allows specifying a label allowlist for all reasources globally. 

Instead of adding a new flag, could we extend the syntax of `--metric-labels-allowlist` to accept `*` as the resource name? You can then use `--metric-labels-allowlist=*=[label1,label2]` to achieve the desired outcome.

**Comment by sherifabdlnaby** [comment]
> Thanks for this contribution. If I understand correctly, this PR allows specifying a label allowlist for all reasources globally.
> 
> Instead of adding a new flag, could we extend the syntax of `--metric-labels-allowlist` to accept `*` as the resource name? You can then use `--metric-labels-allowlist=*=[label1,label2]` to achieve the desired outcome.

@fpetkovski Not really, allow me to clarify.

`--metric-labels-allowlist` whitelists some K8S Labels to be added to the `kube_<resource>_label` metric as Prometheus Labels. Having the K8S Labels in Prometheus allows you to Join with other metrics and use labels to filter for resources.

Example:
Get all `kube_deployment_spec_replicas` of all deployments with the label `app.kubernetes.io/name=emojiapp`
```
kube_deployment_spec_replicas{} * on (deployment) group_left(label_app_kubernetes_io_name) kube_deployment_labels{ label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
```

## What the PR Suggests 
White list some labels to be added to each individual metric so you don't need the complex joins. The Post above specify how complex joins can get.

Using `--per-metric-labels-allowlist` we can whitelist some K8S Labels that will be added **to each individual metric series exported by KSM**, so to achieve a similar query for `kube_deployment_spec_replicas` of all deployments with the label `app.kubernetes.io/name=emojiapp` you can simply do 
```
kube_deployment_spec_replicas{label_app_kubernetes_io_name="emojiapp", namespace=~"emoji"}
```

**Comment by sherifabdlnaby** [comment]
@fpetkovski Any comment on this proposal?

**Comment by sherifabdlnaby** [comment]
I created ticket #1758 to open a discussion for this PR

**Comment by partymaker-py** [comment]
This feature is super useful for me now. Please, merge and release it

**Pull request closed without merging**